Remove old model from the MailPoet\Subscription\Manage class [MAILPOET-4350]

lysyjan opened the pull request:
[MAILPOET-4350]

_How to test_
Verify that the form on the Manage page works properly.

[MAILPOET-4350]: https://mailpoet.atlassian.net/browse/MAILPOET-4350?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

lysyjan committed:
Remove old model from Manage

[MAILPOET-4350]

johnolek commented on mailpoet/lib/Subscription/Manage.php:
```diff
@@ -132,34 +135,39 @@ public function onSave() {
+      $segment = $subscriberSegment->getSegment();
+      if (!$segment) {
+        $this->subscriberSegmentRepository->remove($subscriberSegment);
+        $subscriber->getSubscriberSegments()->removeElement($subscriberSegment);
```
Is this necessary? I'm not very familiar with this code so I'm just trying to wrap my head around it. It seems like when we want to use the subscriber segments again we fetch them from the repository down on L168:171.

If the concern is about saving the subscriber with invalid segments it seems like [this code](https://github.com/mailpoet/mailpoet/blob/9867b8c7ad3e822836ebfca8df592c7a29debfc0/mailpoet/lib/Entities/SubscriberEntity.php#L524-L524) should take care of it.

I definitely don't mind having this in here if it's not strictly necessary! It's good to know that `$subscriber->getSubscriberSegments()` will be correct after this `foreach` loop. Again I just want to make sure I understand what's going on.

lysyjan replied:
Good point. I didn't realize that we have a code that solves it. I will remove it.

johnolek commented on mailpoet/tests/integration/Subscription/ManageTest.php:
```diff
@@ -64,59 +79,65 @@ public function testItDoesntRemoveHiddenSegmentsAndCanResubscribe() {
       $this->settings,
       $this->diContainer->get(NewSubscriberNotificationMailer::class),
       $this->diContainer->get(WelcomeScheduler::class),
+      $this->diContainer->get(CustomFieldsRepository::class),
```
I think that this could all be replaced with a helper I added to the codebase a little while ago and I'm just curious to know your opinion on it. With the helper I think it would look like this:

```php
$manage = $this->getServiceWithOverrides(Manage::class, [
  'urlHelper' => Stub::make(UrlHelper::class, [
    'redirectBack' => null,
  ]),
  'fieldNameObfuscator' => Stub::make(FieldNameObfuscator::class, [
    'deobfuscateFormPayload' => function($data) {
      return $data;
    },
  ]),
  'linkTokens' => Stub::make(LinkTokens::class, [
    'verifyToken' => function($token) {
      return true;
    },
  ]),
  'settings' => $this->settings
]);
```

Do you have a preference for one approach vs the other?

Definitely not a required change!

lysyjan replied:
We can simplify the code when it's possible. This pull request isn't big, so I would replace it with your new method.

johnolek commented on mailpoet/tests/integration/Subscription/ManageTest.php:
```diff
@@ -64,59 +79,65 @@ public function testItDoesntRemoveHiddenSegmentsAndCanResubscribe() {
-    });
-    expect($subscriber->status)->equals(Subscriber::STATUS_SUBSCRIBED);
+    $subscriber = $this->subscribersRepository->findOneById($this->subscriber->getId());
+    if (!$subscriber) throw new InvalidStateException();
```
It feels a little strange to use an exception like this in a test. I'm assuming this is for PHPStan?

It would feel a little better to me to use something like `assert($subscriber instanceof SubscriberEntity);` which I think PHPStan understands, but it's not a big deal one way or another.

lysyjan replied:
If I am not wrong, we want to avoid using the `assert` function. We have [a ticket](https://mailpoet.atlassian.net/browse/MAILPOET-4258) for that. But you are right that the exception looks a bit weird here, so I would change it to `$this->assertInstanceOf(SubscriberEntity::class, $subscriber);`. Do you agree with it?

johnolek replied:
@lysyjan I was familiar with wanting to remove `assert` from the codebase but I didn't realize that applied to our test code as well. Thanks for pointing that out!

The reason I suggested `assert` is because I tried `expect($subscriber)->isInstanceOf(SubscriberEntity::class);`, which PHPStan didn't seem to understand. If PHPStan does understand `$this->assertInstanceOf(SubscriberEntity::class, $subscriber);` I would definitely prefer that!

johnolek reviewed (dismissed):
@lysyjan this looks good to me overall! I had just one question about something I wasn't sure I was understanding. I left a couple other comments about minor things, nothing substantial as far as the PR goes.

I'm marking this approved because it looks good from what I can tell, but I'm going to leave it assigned to you in case you'd like to make any changes.

lysyjan committed:
Remove redundant code removing relation

Removing inconsistent relations is solved in the class SubscriberEntity by the preFlush
[MAILPOET-4350]

lysyjan committed:
Improve integration ManageTest class

[MAILPOET-4350]

johnolek reviewed (approved):
Thanks for making those tweaks!
